### PR TITLE
ajax polling ch.swisstopo-vd.amtliche-vermessung

### PR DESCRIPTION
**Pull request by ltclm**
[Testlink](https://mf-geoadmin3.dev.bgdi.ch/?topic=ech&lang=de&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo-vd.amtliche-vermessung&X=242163.71&Y=621206.58&zoom=6&api_url=%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fdev_ltclm_av_opendata)

### CODE REVIEW

**Comment by ltclm** [comment]
@davidoesch, @hpchrist 
geodata01.admin.ch is not supporting https, could you please ask to activate it.
polling the pdf resource is working with [http](http://mf-chsdi3.dev.bgdi.ch/dev_ltclm_av_opendata/rest/services/all/MapServer/ch.swisstopo-vd.amtliche-vermessung/11/iframeHtmlPopup?lang=de&geometry=624923.0641782399,243370.26669311995), it is not working with [https](https://mf-chsdi3.dev.bgdi.ch/dev_ltclm_av_opendata/rest/services/all/MapServer/ch.swisstopo-vd.amtliche-vermessung/11/iframeHtmlPopup?lang=de&geometry=624923.0641782399,243370.26669311995)

```bash
$ curl -k -I "https://geodata01.admin.ch/download/860936453487269/egris_cadastre.pdf"
HTTP/1.1 200 OK
Server: proxy.admin.ch
Date: Tue, 15 Nov 2016 06:44:25 GMT
Via: 1.1 proxya-bvcol.admin.ch

curl: (35) error:14082174:SSL routines:SSL3_CHECK_CERT_AND_ALGORITHM:dh key too small
```

**Comment by davidoesch** [comment]
@hpchrist can you take care of it?

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 1]
```diff
@@ -31,23 +31,25 @@ $(document).ready(function() {
...
+function stopPolling() {
+    clearTimeout(t);
+}
+
 $('#pdf_download').html('waiting for pdf...');
 (function next() {
     if (counter++ > maxLoops) return;
-
-    setTimeout(function() {
+    t = setTimeout(function() {
```
Comment (oterral) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: Missing var declaration?

**Comment by ltclm** [comment]
thanks to @loicgasser input i have
* moved the polling from iframe to the extended tooltip
* forced http requests for the extended tooltip window and the pdf polling, https is currently not supported by geodata01.admin.ch

@davidoesch 
the first version is now working, please use this [testlink](https://mf-geoadmin3.dev.bgdi.ch/?topic=ech&lang=de&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo-vd.amtliche-vermessung&X=242163.71&Y=621206.58&zoom=6&api_url=%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fdev_ltclm_av_opendata)

**Comment by davidoesch** [comment]
- keep it in tooltip not extended tooltip (as in the poc of Pauline... When waiting for the polling : indicate generating link to PDF as in Pauline's poc) since we also have to show where no PDF download is available.
- on mobile: see this movie https://goo.gl/photos/WPxyGcdcSd3JT8aY6 it shows the loading circle always even when the download started.
- there will also be the need to add the link to the download of the shp / ili file in the tooltip (see spec and poc of Pauline)

So you need to start polling as soon as the user clicks on the map ... Not when starting the tooltip.
We can discuss this on Wed via phone of needed

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 2]
```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
+
+<%inherit file="base.mako"/>
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: you could inherit the other cadastral webmap template or make this one inherit the other one. (like it's done with base.mako hereabove)

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        <tr><td class="cell-left">${_('link to canton geoportal')}</td><td><a href="https://sitonline.vs.ch/urbanisation/mo/${fallbackLang}/" target="_blnk">VS</a></td></br>
+    % elif c['attributes']['ak'] == 'FL':
+        <tr><td class="cell-left">${_('link to geoportal')}</td><td><a href="http://geodaten.llv.li/geoshop/public.html?zoombox=${c['bbox'][0]},${c['bbox'][1]},${c['bbox'][2]},${c['bbox'][3]}" target="_blank">${_('FL')}</a></td></tr>
+    % else:
+        <tr><td class="cell-left">${_('link to canton geoportal')}</td><td>${_('Canton has provided no link to portal')}</td></tr>
+    % endif
+        <tr>
+            <td class="cell-left">${_('ch.kantone.cadastralwebmap-farbe.plan')}</td><td>
+<a href="${'http:%s' % h.make_agnostic(request.route_url('extendedHtmlPopup', map=topic, layerId=c['layerBodId'], featureId=str(c['featureId'])))}?lang=${lang}&pdf_url=${pdf_url}" target="_blank">
+                ${_('zusatzinfo')}&nbsp;<img src="${h.versioned(request.static_url('chsdi:static/images/ico_extern.gif'))}" />
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: I would replace the `zusatzinfo` text.

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako (hunk 2 above)
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: Instead you could inherit the other cadastral webmap mako or make the other one inherit this one.

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+<%
+    import requests
+    pdf_url = context.get('request').params.get('pdf_url')
+    r = requests.get(pdf_url, verify=False)
+    pdf_url = r.text.strip()
+
+    if r.status_code == 200:
+        pdf = True
+    elif pdf_url == 'no object found' :
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: Are you sure here... I would simply use a else statement.

File: chsdi/models/vector/stopo.py [hunk 5]
```diff
@@ -578,6 +578,20 @@ class CadastralWebMap(Base, Vector):
...
 
+class CadastralWebMapOpenData(Base, Vector):
+    __tablename__ = 'swissboundaries_kantone'
+    __table_args__ = ({'schema': 'tlm', 'autoload': False, 'extend_existing': True})
+    __template__ = 'templates/htmlpopup/cadastralwebmap_opendata.mako'
+    __bodId__ = 'ch.swisstopo-vd.amtliche-vermessung'
+    __label__ = 'ak'
+    __extended_info__ = True
+    id = Column('kantonsnr', Integer, primary_key=True)
+    ak = Column('ak', Text)
```
Comment (loicgasser) [inline chsdi/models/vector/stopo.py]: Unicode please

**Comment by davidoesch** [comment]
Are we gona have 2 cadastral web maps?
- one is the cwm: nothing changes.(Pauline has made the test code on this one)
- open-plan is a new layer served by the same wms server from geoshop and it is av data : @ltclm you have to ask Ibele for the wms layer name

**Comment by loicgasser** [comment]
> keep it in tooltip not extended tooltip (as in the poc of Pauline... When waiting for the polling : indicate generating link to PDF as in Pauline's poc) since we also have to show where no PDF download is available.

The major drawback with that is that you will generate a pdf every time you click in the map.
You have to take that into consideration.
Performing the first request that return a pdf URL will generate the PDF in the background even if never used later. I think we may make their server crash pretty easy.
Also, we just fixed a endless polling issue with haltestellen layer. An iframe within a regular tooltip is always suboptimal in my opinion.

> on mobile: see this movie https://goo.gl/photos/WPxyGcdcSd3JT8aY6 it shows the loading circle always even when the download started.

This can be changed to a back link to geoadmin. It's still WIP.

**Comment by davidoesch** [comment]
Pdf generation: why not just poll if the perimeter will be have a PDF? (This can be done with a get feature request on wms, since it is limited on Canton granularity) Then show pdf available or not available: only if clicking on  PDF available -> call for the generation of the PDF.

Possible?

**Comment by loicgasser** [comment]
> Pdf generation: why not just poll if the perimeter will be have a PDF? (This can be done with a get feature request on wms, since it is limited on Canton granularity) Then show pdf available or not available: only if clicking on PDF available -> call for the generation of the PDF.

Yes sure, I would take this approach if it is based on a per canton basis. We need to extend or cadatral web map table somehow...

**Comment by davidoesch** [comment]
It's just: Expert Tooltips are a nuisance - no one will click on additional info. And here this is the USP of the whole dataset :pdf and download. I even doubt that people click on a raster map at all

**Comment by gjn** [comment]
Let's KISS. And optimize if usage is high.
So, I'd vote for least effort that creates an acceptable solution.
Am 15.11.2016 6:04 nachm. schrieb David Oesch <notifications@github.com>:It's just: Expert Tooltips are a nuisance - no one will click on additional info. And here this is the USP of the whole dataset :pdf and download. I even doubt that people click on a raster map at all

**Comment by gjn** [comment]
2 additional points:
- i like iframe approach as it's good encapsulation and assures it works on it's own
- don't like 2 layers for basically same data...
Am 15.11.2016 5:15 nachm. schrieb David Oesch <notifications@github.com>:keep it in tooltip not extended tooltip (as in the poc of Pauline... When waiting for the polling : indicate generating link to PDF as in Pauline's poc) since we also have to show where no PDF download is available.on mobile: see this movie https://goo.gl/photos/WPxyGcdcSd3JT8aY6 it shows the loading circle always even when the download started.there will also be the need to add the link to the download of the shp / ili file in the tooltip (see spec and poc of Pauline)

So you need to start polling as soon as the user clicks on the map ... Not when starting the tooltip.
We can discuss this on Wed via phone of needed

**Comment by davidoesch** [comment]
- KISS for the user: if hide the key feature in expert tooltips- it will not be used. That is a sad fact. What can be an option is based on cantonal level we have a link to a PDF which when clicked opens a new tab .. Served by geodata... There were can have the waiting wheel.
Different dataset: the layer here is the geobasisdaten layer AV. This is legally binding . Cwm not.

**Comment by ltclm** [comment]
@davidoesch 
* polling from iframe within the default tooltip won't work until the geoshop server is supporting https with a valid certificate (@hpchrist has ordered this change today)
* what do you mean with ili/shp download, there is nothing about that in the poc of pauline which i guess is this one [1]
* pdf creation / polling will only be available for the cantons in this layer right **ch.swisstopo-vd.amtliche-vermessung** ? 
* How and where can we get the list of cantons which are opendata? We will have to create a view and activate the tooltip on this cantons only.
* tooltip of cadastral web map won't change ?

In my eyes we should keep this polling stuff in the extended tooltip and not slow down the standard tooltip with javascript and iframe gimmicks.


[1] https://github.com/geoadmin/mf-chsdi3/pull/1996

**Comment by davidoesch** [comment]
- what do you mean with ili/shp download, there is nothing about that in the poc of pauline which i guess is this one [1]
-> Pauline had it once partly set up .Spec in https://redmine.prod.bgdi.ch/issues/6154 

 

> Special Tooltip basing on a service that gives back a link to the data zipfile (2 fields for versions ITF and SHP) if in [1] there is a file for the BFS nr of the click coordinate and on the PDF which are delivered in  see also DIF
> [1] data.geo.admin.ch/ch.swisstopo-vd.amtliche-vermessung/DM01AVCH24D/ 

from the DIF
image


- pdf creation / polling will only be available for the cantons in this layer right ch.swisstopo-vd.amtliche-vermessung ?

-> Yes

- How and where can we get the list of cantons which are opendata? We will have to create a view and activate the tooltip on this cantons only.

-> Ibele will activate the cantons in the WMS.  we have to work with the wms (doubel check with Ibele if there can be done a getFeature to retrieve the perimeter for valid datasets). another way would be: we extend the view which we have in CWM which contains the URL /link to cantonal portal. but this gets to complicated

- tooltip of cadastral web map won't change ?

-> for the go live on 1.12.2016: no . And I insist not to do so with Ibele

- In my eyes we should keep this polling stuff in the extended tooltip and not slow down the standard tooltip with javascript and iframe gimmicks.

-> open a new Tab/Window when the polling starts (but we discussed this a year ago an jeiz/moc had some arguments we don't do it in a new tab/window)

**Comment by gjn** [comment]
Ok, why not have 'link to pdf' in standard tooltip. When clicked, it opens extended tooltip in new window where polling is started, pdf created and auto-downloaded when complete?
No polling on map.geo side, yet diredtly visible link to possible pdf.
Am 15.11.2016 8:13 nachm. schrieb David Oesch <notifications@github.com>:KISS for the user: if hide the key feature in expert tooltips- it will not be used. That is a sad fact. What can be an option is based on cantonal level we have a link to a PDF which when clicked opens a new tab .. Served by geodata... There were can have the waiting wheel.
Different dataset: the layer here is the geobasisdaten layer AV. This is legally binding . Cwm not.

**Comment by davidoesch** [comment]
The winner is: @gjn . This is how we do it. 
The only catch to solve here: we need to know for the normal tooltip of for this coordinate a PDF *can* be generated. So, either we do this with a predefined view pet Canton, or with a simple get feature call on the wms. If there is pdf/download shp/itf available: then we show a "-" in the corresponding col of the tooltip. So we don't need a "Zusatzinformationen" Link in the normal tooltip.
All Clear? If not let's solve this 2morrow with a short phone call

**Comment by ltclm** [comment]
thanks for your feedback so far,
a new version is now ready.

tooltips are now limited to the covered area, there is a table stopo.vd.av_opendata with the list of opendata cantons.
polling is done in the extended tooltips only
translations have been taken from data integration formular
redundant template code has been reduced with includes
geodata01 is now supporting https

@davidoesch 
the background of the covered area in the geoshop wms should be white, it is transparent now. can you ask inforgrips to change this?
* Opendata layer in the following link is the pure geoshop wms.
* GB-Plan layer is the cascaded layer

[Testlink](https://mf-geoadmin3.dev.bgdi.ch/?topic=ech&lang=de&bgLayer=ch.swisstopo.swissimage&layers=ch.swisstopo-vd.amtliche-vermessung,WMS%7C%7COPENDATA%7C%7Chttp:%2F%2Fwms.cadastralwebmap.ch%2Fo9uoulnl77h%2FWMS%7C%7Copendata%7C%7C1.1.1&X=223400.87&Y=618721.89&zoom=10&api_url=%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fdev_ltclm_av_opendata&layers_opacity=0.4,1&layers_visibility=true,false)

**Comment by gjn** [comment]
I like it!

**Comment by davidoesch** [comment]
nicely done! i like

observation:
- on desktop it is fine, on mobile the loading wheel just don't stop and no download of PDF. yes cache cleaned restarted browser etc...

Improvement:
Since the expert tooltip in this case here, all the yellow parts below do not make sense (user wont be able to click on any of the links or go back to the tooltip)
image

-> only make a pure html site containing:

EN: Please wait until your document is ready for download ... 
DE: Bitte warten Sie, während das Dokument zum Download vorbereitet wird ...
FR: Veuillez patienter pendant la préparation de ce document pour le téléchargement...
IT: Attendere mentre il documento viene preparato per la Download...

enter a <p> and <br> more between the text and the spinning wheel.

if not a pure html is possible: then just add the translations above.

Ill inform Ibele and tell him as well to test.

**Comment by davidoesch** [comment]
and here an issue:
when opening this link
https://mf-geoadmin3.dev.bgdi.ch/?topic=ech&lang=de&bgLayer=voidLayer&layers=ch.swisstopo-vd.amtliche-vermessung,WMS%7C%7COPENDATA%7C%7Chttp:%2F%2Fwms.cadastralwebmap.ch%2Fo9uoulnl77h%2FWMS%7C%7Copendata%7C%7C1.1.1&X=241602.00&Y=629524.56&zoom=13&api_url=%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fdev_ltclm_av_opendata&layers_opacity=0.4,1&layers_visibility=true,false

clickling on a parzelle
 I get notification that a PDF is available
image

but if i want to download it:
I get
image

So I assume:
is this based on the polling result? -> clm?
even if cantonwide PDF data should be available there are some exceptions?-> Will ask ibele

the idea would have been: we show a "-" for perimeters where no PDF are availbale in the tooltip

**Comment by ltclm** [comment]
the issue mentioned above was due to a typo in the get parameter, it should work fine again now.
Will change the extended tooltip layout tomorrow.

**Comment by davidoesch** [comment]
Works now like a charm... And thanks for the translation and clean up. 
Question: error handling. What will be the response if the polling is not successful/server is down? "Pdf not available"  as above makes sense

**Comment by ltclm** [comment]
@davidoesch 
yes the default message will be "pdf not available" when
* pdf does not exist ([1] status not 200)
* the service is down (generated pdf url is wrong)
* pdf is not ready within 50 loops (+/- 50 seconds)

[1] http://geodata01.admin.ch/order/jPqrueQazrt/av_pdf.igs?pos=%s/%s

**Comment by ltclm** [comment]
@davidoesch 
* tooltip geometry source is municipalities now
* shp and interlis link have been added

**Comment by davidoesch** [comment]
Nice job-Perfect ... Just add "Gemeinde"in brackets  as well to the itf download. As soon as German changed the background: we are ready for AP. Cc @mariokeusen

**Comment by gjn** [comment]
I like it as well.

Could you also squash the commits?

**Comment by loicgasser** [comment]
we can when merging now

File: chsdi/static/doc/source/releasenotes/index.rst [hunk 6]
```diff
@@ -51,12 +50,15 @@ Geodata
 +--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
 | Update | `Energy-Regions <//map.geo.admin.ch/?layers=ch.bfe.energiestaedte-energieregionen>`__ (ch.bfe.energiestaedte-energieregionen)                               |
 +--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
+<<<<<<< HEAD
```
Comment (loicgasser) [inline chsdi/static/doc/source/releasenotes/index.rst]: rebase error

File: chsdi/static/doc/source/releasenotes/index.rst [hunk 7]
```diff
@@ -51,12 +50,15 @@ Geodata
...
 | Update | `Low distortion area <//map.geo.admin.ch/?layers=ch.swisstopo-vd.spannungsarme-gebiete>`__ (ch.swisstopo-vd.spannungsarme-gebiete)                          |
 +--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
 | Update | `Designated wildlife areas <//map.geo.admin.ch/?layers=ch.bafu.wrz-wildruhezonen_portal>`__ (ch.bafu.wrz-wildruhezonen_portal)                              |
 +--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
 | Update | `Wildlife reserves <//map.geo.admin.ch/?layers=ch.bafu.wrz-jagdbanngebiete_select>`__ (ch.bafu.wrz-jagdbanngebiete_select)                                  |
 +--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
+=======
+>>>>>>> fix releasenotes label
```
Comment (loicgasser) [inline chsdi/static/doc/source/releasenotes/index.rst]: and here

**Comment by ltclm** [comment]
@davidoesch i have a updated the translation as desired
commits have been squashed

**Comment by davidoesch** [comment]
Thanks Marcel... Really impressive job

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako (hunk 2 above)
Comment (ltclm) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: i have tried to use inherit but i did not make it. so for now it is done with an include.

Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: An include is good as well at least we don't have to change these URLs in 2 places

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  c['bboxlv95'] = [2000000 + c['bbox'][0], 1000000 + c['bbox'][1], 2000000 + c['bbox'][2], 1000000 + c['bbox'][3]]
+  c['scale']  = getScale(params.imageDisplay, params.mapExtent)
+  topic = request.matchdict.get('map')
+  baseUrl = request.registry.settings['api_url']
+  coord = request.params.get('coord').split(',')
+  lat = coord[0]
+  lon = coord[1]
+  pdf_url = "%s://geodata01.admin.ch/order/jPqrueQazrt/av_pdf.igs?pos=%s/%s" % (protocol, lat, lon)
+  shp_url = "%s://data.geo.admin.ch/ch.swisstopo-vd.amtliche-vermessung/DM01AVCH24D/SHP/%s/%s.zip" % (protocol, c['attributes']['ak'],c['featureId'])
+  itf_url = "%s://data.geo.admin.ch/ch.swisstopo-vd.amtliche-vermessung/DM01AVCH24D/ITF/%s/%s.zip" % (protocol, c['attributes']['ak'],c['featureId'])
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: You should be able to get the hostname from request.registry.settings

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            visibility: hidden;
+        }
+
+        div#pdf_download * {
+            visibility: visible;
+        }
+
+    </style>
+
+% if pdf == True:
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: If pdf is probably sufficient, minor

**Comment by ltclm** [comment]
thanks @loicgasser  i have addressed your comments.

**Comment by loicgasser** [comment]
Great !

**Pull request merged**